src/librustc_mir/interpret/place.rs

```diff
@@ -1082,17 +1082,21 @@ where
                 }
                 if variant_index != dataful_variant {
                     let variants_start = niche_variants.start().as_u32();
-                    let variant_index_relative = variant_index.as_u32()
-                        .checked_sub(variants_start)
-                        .expect("overflow computing relative variant idx");
+
+                    let (variant_index_relative, op) = if variant_index.as_u32() >= variants_start {
+                        (variant_index.as_u32() - variants_start, mir::BinOp::Add)
+                    } else {
+                        (variants_start - variant_index.as_u32(), mir::BinOp::Sub)
+                    };
+
                     // We need to use machine arithmetic when taking into account `niche_start`:
                     // discr_val = variant_index_relative + niche_start_val
                     let discr_layout = self.layout_of(discr_layout.value.to_int_ty(*self.tcx))?;
                     let niche_start_val = ImmTy::from_uint(niche_start, discr_layout);
                     let variant_index_relative_val =
                         ImmTy::from_uint(variant_index_relative, discr_layout);
                     let discr_val = self.binary_op(
-                        mir::BinOp::Add,
+                        op,
```

Comment:
This doesn't seem right to me... in the 2nd branch above, what happens (compared to earlier) is that `variant_index_relative` is negative, but here by changing the operator you are negating the *second* operand, `niche_start_val`. So the entire result is the negation of what it used to be. Don't you also have to swap the operands?

Either way, this needs a *detailed* comment why both branches above are correct. Also we should understand why LLVM codegen doesn't need a branch like this.